### ISSUE TITLE
Training on Other Datasets

### ISSUE THREAD

**Issue by fcr3** [ManyiLi12345/GRAINS#8]
What is required for GRAINS to be trained on other datasets? Since suncg is down, it's not clear as to how we can ingest data into GRAINS. Can you provide some metadata for how we should organize our own custom dataset?

**Comment by ManyiLi12345** [comment]
The "4-training" of our code is the core part for the training, and most of the rest is developed to parse the SUNCG dataset. We suggest to process the other dataset in the form as described in our paper, and then directly call the training code. Please check the sample training data including two rooms [here](https://github.com/ManyiLi12345/GRAINS/tree/master/4-training/sample_trainingdata_with_two_rooms).

**Comment by fcr3** [comment]
Thanks for the response! Is there a metadata README describing the training sample data? I saw that the relpos1.mat fits the 28-dimensional vector. But what about the other files? Are they necessary for training?

**Comment by fcr3** [comment]
I also just noticed in module 2 that updateOBB is not defined. Can you provide the code for this script? Thanks :)

**Comment by ManyiLi12345** [comment]
> I also just noticed in module 2 that updateOBB is not defined. Can you provide the code for this script? Thanks :)

Please find the script updateOBB.m [here](https://github.com/ManyiLi12345/GRAINS/blob/master/vistools/updateOBB.m).

**Comment by ManyiLi12345** [comment]
> Thanks for the response! Is there a metadata README describing the training sample data? I saw that the relpos1.mat fits the 28-dimensional vector. But what about the other files? Are they necessary for training?

The relative positions are always 28-dimensional vectors. The format is described in our paper, so we didn't provide a README about the training data format here. With our network architecture, it's necessary to have all the position vectors in the same length.

**Comment by fcr3** [comment]
> > Thanks for the response! Is there a metadata README describing the training sample data? I saw that the relpos1.mat fits the 28-dimensional vector. But what about the other files? Are they necessary for training?
> 
> The relative positions are always 28-dimensional vectors. The format is described in our paper, so we didn't provide a README about the training data format here. With our network architecture, it's necessary to have all the position vectors in the same length.

@ManyiLi12345 Sorry I don't think my question was answered. After diving into the training code, I understand that the ops1.mat guides the tree building process and from reading the paper the relpos1.mat is the 28D vector encodings, but I am still unclear as to what the boxes1.mat represents.

**Comment by ManyiLi12345** [comment]
> > > Thanks for the response! Is there a metadata README describing the training sample data? I saw that the relpos1.mat fits the 28-dimensional vector. But what about the other files? Are they necessary for training?
> > 
> > 
> > The relative positions are always 28-dimensional vectors. The format is described in our paper, so we didn't provide a README about the training data format here. With our network architecture, it's necessary to have all the position vectors in the same length.
> 
> @ManyiLi12345 Sorry I don't think my question was answered. After diving into the training code, I understand that the ops1.mat guides the tree building process and from reading the paper the relpos1.mat is the 28D vector encodings, but I am still unclear as to what the boxes1.mat represents.

Oh, sorry I misunderstood your questions. In short, the objects' sizes and labels are stored in boxes1.mat. The positions are decomposed into relative positions between tree nodes, stored in the relpos1.mat. The script 5-reconVAE/main_recon.m visualizes the generated scenes in the same format as the training data.

For example, we consider objects from M=20 categories for bedrooms. In the training data, each room has at most K=20 object instances. For one bedroom with 14 objects, it corresponds to a (3+M)*K sub-matrix in the boxes1.mat. The first 14 columns are for each object and the rest K-14 columns are padded with zeros. Each column concatenates a 3D vector [s1,s2,s3] for the sizes along the "front", "up" and the last direction, and a one-hot vector for its label.

**Comment by fcr3** [comment]
@ManyiLi12345  Okay this makes sense. For the one-hot encoded label, does this correspond to the same order as the labelsets folder in `0-data`?

For instance, a wall is the 20th (last) item in the bedroom labelset, so I would expect that the one-hot encoded vector would be many zeros with the last spot being a 1 in `boxes1.mat`?

**Comment by ManyiLi12345** [comment]
> @ManyiLi12345 Okay this makes sense. For the one-hot encoded label, does this correspond to the same order as the labelsets folder in `0-data`?
> 
> For instance, a wall is the 20th (last) item in the bedroom labelset, so I would expect that the one-hot encoded vector would be many zeros with the last spot being a 1 in `boxes1.mat`?

Yes, that's right.

**Comment by fcr3** [comment]
@ManyiLi12345  Thanks for answer my questions btw! I was wondering if one item can be attached to a node. After preprocessing my data, I've seen that one sofa is closer to one wall than the others, but no other items are closer to this wall than the others. So it seems that the graph will contain a wall node with only child. Is this allowed?

**Comment by fcr3** [comment]
> @ManyiLi12345 Thanks for answer my questions btw! I was wondering if one item can be attached to a node. After preprocessing my data, I've seen that one sofa is closer to one wall than the others, but no other items are closer to this wall than the others. So it seems that the graph will contain a wall node with only child. Is this allowed?

Nevermind, the wallnode encompasses the wall box and a standalone item.
